src/main/java/fea/storage/Storage.java

```diff
@@ -0,0 +1,127 @@
+
+    public Storage(String filePath) {
+        this.filePath = Paths.get(filePath);
+    }
+    /**
+     * Loads the tasks from the data file into the task list.
+     *
+     * @return ArrayList Returns an ArrayList containing the tasks from the data file.
+     * @throws FeaException If the data file could not be found.
+     */
+    public ArrayList<Task> loadTasks() throws FeaException {
+        File file = this.filePath.toFile();
+        if (!file.exists()) {
+            throw new FeaException(FILE_NOT_FOUND);
+        }
+        ArrayList<Task> loadedList = new ArrayList<>();
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm");
+        try (Scanner scanner = new Scanner(file)) {
+            while (scanner.hasNextLine()) {
+                String[] args = scanner.nextLine().split(" \\| ");
+                checkTaskFormat(args[0], args[1]);
+                switch (args[0]) {
+                case "T":
+                    Todo newTodo = new Todo(args[2]);
+                    if (args[1].equals("1")) {
+                        newTodo.toggleMark();
+                    }
+                    loadedList.add(newTodo);
+                    break;
+                case "D":
+                    Deadline newDeadline = new Deadline(args[2], LocalDateTime.parse(args[3], formatter));
+                    if (args[1].equals("1")) {
+                        newDeadline.toggleMark();
+                    }
+                    loadedList.add(newDeadline);
+                    break;
+                case "E":
+                    Event newEvent = new Event(args[2], LocalDateTime.parse(args[3], formatter),
+                            LocalDateTime.parse(args[4], formatter));
+                    if (args[1].equals("1")) {
+                        newEvent.toggleMark();
+                    }
+                    loadedList.add(newEvent);
+                    break;
+                default:
+                    break;
+                }
+            }
+        } catch (FileNotFoundException e) {
+            throw new FeaException(FILE_NOT_FOUND);
+        }
+        return loadedList;
+    }
+    /**
+     * Saves the tasks in the tasklist into the data file.
+     *
+     * @param tasks The task list converted to an ArrayList to be saved.
+     * @throws FeaException If the data file could not be found or there was an error creating the data file.
+     */
+    public void saveTasks(ArrayList<Task> tasks) throws FeaException {
+        File file = filePath.toFile();
+        if (!file.exists()) {
+            file.getParentFile().mkdirs();
+            try {
+                file.createNewFile();
+            } catch (IOException e) {
+                throw new FeaException(fileNotCreated);
+            }
+        }
+
+        try (FileWriter fileWriter = new FileWriter(file)) {
+            for (Task task : tasks) {
+                int mark = task.getMark().equals('X') ? 1 : 0;
```

Comment:
If I'm not wrong, using instanceof is usually a bad practice as it misses out on OOP design, maybe you could try creating/using the respective task class's functions to do the checks instead?

Reply:
Other that that, very clean way of saving tasks